src/broadcaster.rs

```diff
@@ -0,0 +1,192 @@
+                // nothing new to report
+                status.waker = Some(cx.waker().clone());
+                Async::Pending
+            }
+        }
+    }
+}
+
+impl<A: Signal> Clone for Broadcaster<A>
+    where <A as Signal>::Item: Clone {
+    fn clone(&self) -> Self {
+
+        let new_status = Arc::new(Mutex::new(BroadcasterStatus {
+            has_changed: self.state.status.lock().unwrap().has_changed,
+            waker: None
+        }));
+
+        let shared_state = self.state.shared_state.lock().unwrap();
+        shared_state.notifier.targets.lock().unwrap().push(Arc::downgrade(&new_status));
+
+        let new_state = Arc::new(BroadcasterState {
+            status: new_status,
+            shared_state: self.state.shared_state.clone()
+        });
+
+        Self {
+            state: new_state,
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
+
+// State for a single broadcaster instance.
+//
+// It's split in this way because the BroadcasterSharedState also needs to access the status
+// (e.g. to notify of changes).
+struct BroadcasterState<A: Signal> {
+    status: Arc<Mutex<BroadcasterStatus>>,
+    shared_state: Arc<Mutex<BroadcasterSharedState<A>>>,
+}
+
+// ---------------------------------------------------------------------------------------------
+
+struct BroadcasterStatus {
+    has_changed: bool,
+    waker: Option<Waker>
+}
+
+// ---------------------------------------------------------------------------------------------
+
+// Shared state underpinning a Cloned set of broadcasters
+struct BroadcasterSharedState<A: Signal> {
+    signal: A,
+    value: Option<A::Item>,
+    notifier: Arc<BroadcasterNotifier>
+}
+
+impl<A: Signal> BroadcasterSharedState<A>
+        where <A as Signal>::Item: Clone {
+
+    // Poll the underlying signal for changes, giving it a BroadcasterWaker to wake in
+    // the future if it is in Pending state.
+    //
+    // Takes self by Arc<Mutex<Self>> because this construct is needed to build a
+    // BroadcasterWaker
+    fn poll_change(&mut self, cx: &mut Context) -> Async<Option<A::Item>> {
+
+        // Set the notifier to be waiting on a poll, and if it wasn't previously, actually
+        // execute that poll.
+        //
+        // TODO: Do we actually need AcqRel guarantee?
+        if !self.notifier.is_waiting.swap(true, Ordering::AcqRel) {
+            let waker = Waker::from(self.notifier.clone());
+            let mut sub_cx = cx.with_waker(&waker);
+
+            if let Async::Ready(value) = self.signal.poll_change(&mut sub_cx) {
+                self.value = value.clone();
+                self.notifier.notify();
+                self.notifier.is_waiting.store(false, Ordering::Release);
+                return Async::Ready(value)
+            }
+        }
+
+        Async::Pending
+    }
+}
+
+// ----------------------------------------------------------------------------------------------
```

Comment:
This goes well over the 80-character limit (which isn't currently hard-enforced, but still).